### PR TITLE
# 535 Enhancements to Adding, Revoking, and Removing Non-Root Certificates

### PR DESCRIPTION
**Pull request by akarabashov**

### CODE REVIEW

File: x/pki/keeper/keeper.go [hunk 1]
```diff
@@ -50,3 +50,29 @@ func (k Keeper) CertificateApprovalsCount(ctx sdk.Context, authKeeper types.Dcla
 func (k Keeper) CertificateRejectApprovalsCount(ctx sdk.Context, authKeeper types.DclauthKeeper) int {
 	return authKeeper.CountAccountsWithRole(ctx, authTypes.Trustee) - k.CertificateApprovalsCount(ctx, authKeeper) + 1
 }
+
+func (k Keeper) EnsureSenderAndOwnerVidMatch(ctx sdk.Context, certificate *types.Certificate, signer string) error {
```
Comment (Abdulbois) [inline x/pki/keeper/keeper.go:54]:
```suggestion
func (k Keeper) VerifyVidMatching(ctx sdk.Context, certificate *types.Certificate, signer string) error {
```

File: x/pki/keeper/msg_server_add_x_509_cert.go [hunk 2]
```diff
@@ -6,13 +6,24 @@ import (
...
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
+	signerAddr, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		return nil, pkitypes.NewErrInvalidAddress(err)
+	}
+
+	// check if signer has vendor role
+	if !k.dclauthKeeper.HasRole(ctx, signerAddr, dclauthtypes.Vendor) {
+		return nil, pkitypes.NewErrUnauthorizedRole("MsgAddNocX509RootCert", dclauthtypes.Vendor)
```
Comment (Abdulbois) [inline x/pki/keeper/msg_server_add_x_509_cert.go:24]:
```suggestion
		return nil, pkitypes.NewErrUnauthorizedRole("MsgAddX509Cert", dclauthtypes.Vendor)
```

File: x/pki/keeper/msg_server_add_x_509_cert.go [hunk 3]
```diff
@@ -50,45 +58,32 @@ func (k msgServer) AddX509Cert(goCtx context.Context, msg *types.MsgAddX509Cert)
...
-		// Only a Vendor associated with this VID can add an intermediate certificate. So `rootVid == childVid == accountVID`
-		// condition must hold
-		if rootVid != childVid || rootVid != accountVID {
-			return nil, pkitypes.NewErrRootCertVidNotEqualToAccountVidOrCertVid(rootVid, accountVID, childVid)
-		}
-		// If added under a non-VID scoped root CA associated with a VID: Intermediate cert must be either VID scoped to the same VID, or non-VID scoped.
-		// Only a Vendor associated with this VID can add an intermediate certificate.
-	} else if childVid != 0 && childVid != accountVID {
-		return nil, pkitypes.NewErrAccountVidNotEqualToCertVid(accountVID, childVid)
+	// Provided certificate, root certificate and account VID must match
```
Comment (Abdulbois) [inline x/pki/keeper/msg_server_add_x_509_cert.go:84]:
```suggestion
	// VID of account must match to VID of root and provided child certificates
```

File: x/pki/keeper/msg_server_add_x_509_cert.go [hunk 4]
```diff
@@ -131,3 +126,49 @@ func (k msgServer) AddX509Cert(goCtx context.Context, msg *types.MsgAddX509Cert)
 
 	return &types.MsgAddX509CertResponse{}, nil
 }
+
+func (k msgServer) ensureCertsAndSenderVidMatch(
```
Comment (Abdulbois) [inline x/pki/keeper/msg_server_add_x_509_cert.go:130]:
```suggestion
func (k msgServer) verifyVidMatching(
```

File: x/pki/keeper/msg_server_revoke_x_509_cert.go [hunk 5]
```diff
@@ -2,50 +2,50 @@ package keeper
...
-	certificates, found := k.GetApprovedCertificates(ctx, msg.Subject, msg.SubjectKeyId)
-	if !found {
+	signerAddr, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		return nil, pkitypes.NewErrInvalidAddress(err)
+	}
+
+	// check if signer has vendor role
+	if !k.dclauthKeeper.HasRole(ctx, signerAddr, dclauthtypes.Vendor) {
+		return nil, pkitypes.NewErrUnauthorizedRole("MsgAddNocX509RootCert", dclauthtypes.Vendor)
```
Comment (Abdulbois) [inline x/pki/keeper/msg_server_revoke_x_509_cert.go:22]:
```suggestion
		return nil, pkitypes.NewErrUnauthorizedRole("MsgRevokeX509Cert", dclauthtypes.Vendor)
```

File: x/pki/keeper/msg_server_remove_x_509_cert.go [hunk 6]
```diff
@@ -6,15 +6,27 @@ import (
...
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
+	signerAddr, err := sdk.AccAddressFromBech32(msg.Signer)
+	if err != nil {
+		return nil, pkitypes.NewErrInvalidAddress(err)
+	}
+
+	// check if signer has vendor role
+	if !k.dclauthKeeper.HasRole(ctx, signerAddr, dclauthtypes.Vendor) {
+		return nil, pkitypes.NewErrUnauthorizedRole("MsgAddNocX509RootCert", dclauthtypes.Vendor)
```
Comment (Abdulbois) [inline x/pki/keeper/msg_server_remove_x_509_cert.go:23]:
```suggestion
		return nil, pkitypes.NewErrUnauthorizedRole("RemoveX509Cert", dclauthtypes.Vendor)
```

File: x/pki/handler_add_non_root_cert_test.go [hunk 7]
```diff
@@ -0,0 +1,636 @@
...
+	vendorAccAddress := GenerateAccAddress()
+	setup.AddAccount(vendorAccAddress, []dclauthtypes.AccountRole{dclauthtypes.Vendor}, testconstants.Vid)
+
+	// Store the NOC certificate
+	nocCertificate := intermediateCertificate(vendorAccAddress)
+	nocCertificate.SerialNumber = testconstants.TestSerialNumber
+	nocCertificate.IsNoc = true
+
+	setup.Keeper.AddApprovedCertificate(setup.Ctx, nocCertificate)
+	// TODO: add the certificate to the ICA store after the store is implemented
```
Comment (Abdulbois) [inline x/pki/handler_add_non_root_cert_test.go:137]: Seems should be replaced with impl

File: x/pki/handler_add_non_root_cert_test.go [hunk 8]
```diff
@@ -0,0 +1,636 @@
...
+		require.Equal(t, testconstants.IntermediateSubjectKeyID, certificate.SubjectKeyId)
+		require.False(t, certificate.IsRoot)
+		require.Equal(t, testconstants.RootSubject, certificate.RootSubject)
+		require.Equal(t, testconstants.RootSubjectKeyID, certificate.RootSubjectKeyId)
+		require.Equal(t, testconstants.IntermediateIssuer, certificate.Issuer)
+		require.Equal(t, testconstants.IntermediateAuthorityKeyID, certificate.AuthorityKeyId)
+	}
+}
+
+func TestHandler_AddX509Cert_ForAbsentDirectParentCert(t *testing.T) {
```
Comment (Abdulbois) [inline x/pki/handler_add_non_root_cert_test.go:228]:
```suggestion
func TestHandler_AddX509Cert_WhenParentCertAbsent(t *testing.T) {
```

File: x/pki/handler_add_non_root_cert_test.go [hunk 9]
```diff
@@ -0,0 +1,636 @@
...
+	vendorAccAddress := GenerateAccAddress()
+	setup.AddAccount(vendorAccAddress, []dclauthtypes.AccountRole{dclauthtypes.Vendor}, testconstants.Vid)
+
+	// add intermediate x509 certificate
+	addX509Cert := types.NewMsgAddX509Cert(vendorAccAddress.String(), testconstants.IntermediateCertPem)
+	_, err := setup.Handler(setup.Ctx, addX509Cert)
+	require.ErrorIs(t, err, pkitypes.ErrInvalidCertificate)
+}
+
+func TestHandler_AddX509Cert_ForFailedCertificateVerification(t *testing.T) {
```
Comment (Abdulbois) [inline x/pki/handler_add_non_root_cert_test.go:240]:
```suggestion
func TestHandler_AddX509Cert_WhenCertificateVerificationFails(t *testing.T) {
```

File: x/pki/handler_test.go [hunk 10]
```diff
@@ -712,1720 +712,580 @@ func TestHandler_AddX509RootCertsBySubjectKeyId(t *testing.T) {
...
 
-	// store root certificate
-	rootCertOptions := createPAACertWithNumericVidOptions()
+	// propose x509 root certificate by `setup.Trustee` and approve by another trustee
+	rootCertOptions := createTestRootCertOptions()
 	proposeAndApproveRootCertificate(setup, setup.Trustee1, rootCertOptions)
 
-	accAddress := GenerateAccAddress()
-	setup.AddAccount(accAddress, []dclauthtypes.AccountRole{dclauthtypes.Vendor}, testconstants.PAACertWithNumericVidVid)
+	// store another trustee
```
Comment (Abdulbois) [inline x/pki/handler_test.go:755]:
```suggestion
	// add another trustee
```

File: x/pki/handler_test.go [hunk 11]
```diff
@@ -712,1720 +712,580 @@ func TestHandler_AddX509RootCertsBySubjectKeyId(t *testing.T) {
...
-	allApprovedCertificates, _ = queryAllApprovedCertificatesBySubjectKeyID(setup, testconstants.IntermediateSubjectKeyID)
-	require.Equal(t, 0, len(allApprovedCertificates))
-	allApprovedCertificates, _ = queryAllApprovedCertificatesBySubjectKeyID(setup, testconstants.LeafSubjectKeyID)
+	require.Equal(t, testconstants.RootCertPem, allRevokedCertificates[1].Certs[0].PemCert)
+	require.Equal(t, testconstants.IntermediateSubject, allRevokedCertificates[2].Subject)
+	require.Equal(t, testconstants.IntermediateSubjectKeyID, allRevokedCertificates[2].SubjectKeyId)
+	require.Equal(t, 1, len(allRevokedCertificates[2].Certs))
+	require.Equal(t, testconstants.IntermediateCertPem, allRevokedCertificates[2].Certs[0].PemCert)
+
+	// check that no certificates stays approved
```
Comment (Abdulbois) [inline x/pki/handler_test.go:1214]:
```suggestion
	// check that approved certs list is empty
```

File: x/pki/handler_add_non_root_cert_test.go (hunk 9 above)
Comment (akarabashov) [reply x/pki/handler_add_non_root_cert_test.go:240]: Let's keep it for consistency with other test cases names

**Review by Abdulbois — APPROVED** [review]

**Pull request merged**